EFCoreRelationshipsPractice/Services/CompanyService.cs

```diff
@@ -18,22 +20,53 @@ public CompanyService(CompanyDbContext companyDbContext)
 
         public async Task<List<CompanyDto>> GetAll()
         {
-            throw new NotImplementedException();
+            var companies = await companyDbContext.Companies
+                .Include(company => company.ProfileEntity)
+                .Include(company => company.Employees)
+                .ToListAsync();
+            return companies.Select(companyEntity => new CompanyDto(companyEntity)).ToList();
         }
 
         public async Task<CompanyDto> GetById(long id)
         {
-            throw new NotImplementedException();
+            var foundCompanyEntity = await companyDbContext.Companies
+                .Include(company => company.ProfileEntity)
+                .Include(company => company.Employees)
+                .FirstAsync(companyEntity => companyEntity.Id == id);
+            return new CompanyDto(foundCompanyEntity);
         }
 
         public async Task<int> AddCompany(CompanyDto companyDto)
         {
-            throw new NotImplementedException();
+            var companyEntity = new CompanyEntity(companyDto);
+            await companyDbContext.Companies.AddRangeAsync(companyEntity);
+            await companyDbContext.SaveChangesAsync();
+            return companyEntity.Id;
         }
 
         public async Task DeleteCompany(int id)
         {
-            throw new NotImplementedException();
+            var foundCompany = await companyDbContext.Companies
+                .Include(company => company.ProfileEntity)
+                .Include(company => company.Employees)
+                .FirstOrDefaultAsync(company => company.Id == id);
+
+            var foundProfile = await companyDbContext.Profiles
+                .FirstOrDefaultAsync(profile => profile.Id == foundCompany.ProfileEntity.Id);
+            companyDbContext.Profiles.Remove(foundProfile);
+
+            foreach (var employeeEntity in foundCompany.Employees)
+            {
+                var foundEmployee = await companyDbContext.Employees.FirstOrDefaultAsync(employee => employee.Id == employeeEntity.Id);
+                if (foundEmployee != null)
+                {
+                    companyDbContext.Employees.Remove(foundEmployee);
+                }
+            }
+
+            companyDbContext.Companies.Remove(foundCompany);
+
+            await companyDbContext.SaveChangesAsync();
```

Comment:
Cool! Manaully implemented full delete logic, searching for all related entities.